drivers: stepper: Fix stepper callbacks when using work_q

jbr7rr opened the pull request:
Fix issue where stepper callbacks were not being called when using work_q. This was due to the steps being counted down before the work_q was rescheduled.

Added additional unit tests to stepper_api to verify the fix. Needed to increase the tolerance to 30% because of the timing of the work_q.

Fixes: https://github.com/zephyrproject-rtos/zephyr/issues/88821

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/native_sim_adi_tmc2209_work_q.overlay:
```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright (c) 2025 Jilay Sandeep Pandya
```
```suggestion
 * Copyright (c) 2025 Your Name
```

jilaypandya commented on tests/drivers/stepper/stepper_api/testcase.yaml:
```diff
@@ -16,6 +16,15 @@ tests:
+      - platform:native_sim/native/64:DTC_OVERLAY_FILE="boards/native_sim_adi_tmc2209_work_q.overlay"
+    extra_configs:
+      - CONFIG_GPIO=y
+      - CONFIG_COUNTER=y
```
```suggestion
```

jilaypandya commented on tests/drivers/stepper/stepper_api/testcase.yaml:
```diff
@@ -25,6 +34,15 @@ tests:
+      - platform:native_sim/native/64:DTC_OVERLAY_FILE="boards/native_sim_allegro_a4979_work_q.overlay"
+    extra_configs:
+      - CONFIG_GPIO=y
+      - CONFIG_COUNTER=y
```
```suggestion
```

jilaypandya commented on tests/drivers/stepper/stepper_api/testcase.yaml:
```diff
@@ -34,6 +52,15 @@ tests:
+      - platform:native_sim/native/64:DTC_OVERLAY_FILE="boards/native_sim_ti_drv84xx_work_q.overlay"
+    extra_configs:
+      - CONFIG_GPIO=y
+      - CONFIG_COUNTER=y
```
```suggestion
```

jilaypandya commented on tests/drivers/stepper/stepper_api/boards/native_sim_allegro_a4979_work_q.overlay:
```diff
@@ -0,0 +1,26 @@
+/*
+ * Copyright (c) 2025 Jilay Sandeep Pandya
```
```suggestion
 * Copyright (c) 2025 Your Name :)
```

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -140,15 +140,47 @@ ZTEST_F(stepper, test_target_position_w_fixed_step_interval)
+{
+	int32_t steps = 20;
+
+	(void)stepper_enable(fixture->dev);
```
```suggestion
```
This is already done in test setup

jilaypandya commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -140,15 +140,47 @@ ZTEST_F(stepper, test_target_position_w_fixed_step_interval)
+{
+	int32_t steps = -20;
+
+	(void)stepper_enable(fixture->dev);
```
same here, this is already done in test setup
```suggestion
```

jilaypandya reviewed (commented):
Thanks for this Patch :)

can you run this script locally in order to fix compliance checks `./scripts/ci/check_compliance.py`.

dipakgmx commented on tests/drivers/stepper/stepper_api/src/main.c:
```diff
@@ -140,15 +140,47 @@ ZTEST_F(stepper, test_target_position_w_fixed_step_interval)
+	/* timeout is set with 30% tolerance */
 	POLL_AND_CHECK_SIGNAL(stepper_signal, stepper_event, STEPPER_EVENT_STEPS_COMPLETED,
-			      K_MSEC(pos * 120));
+			      K_MSEC(pos * 130));
```
Might as well make the 130 a define or a config

jbr7rr replied:
Would adding Kconfig file with the parameter to the test and then adding the parameter to the testcase.yaml be a good solution?

dipakgmx replied:
Yeah, that would be great actually. Thanks.

faxe1008 reviewed (commented):
This should be seperated into two commits.

jbr7rr commented:
> This should be seperated into two commits.

How? Unit test and fix?

jilaypandya commented:
> > This should be seperated into two commits.
> 
> How? Unit test and fix?

Yep, the order would have to be fix and unit-test. Making 130 as a Kconfig could be added to the second commit as well :)

jbr7rr committed:
drivers: stepper: Fix stepper callbacks when using work_q

Fix issue where stepper callbacks were not being called when
using work_q. This was due to the steps being counted down before
the work_q was rescheduled.

Signed-off-by: Josselin Bunt <josselin@sensible.health>

jbr7rr committed:
drivers: stepper: Add unit tests for stepper API using work_q

This commit adds unit tests for the stepper API using work_q scheduler.

Signed-off-by: Josselin Bunt <josselin@sensible.health>
